lib/net.c

```diff
@@ -79,6 +82,171 @@ find_num_numa_nodes(void)
 	return nb_numa_nodes;
 }
 
+static int
+configure_queue(uint8_t port_id, uint16_t queue_id, enum queue_type ty,
+	unsigned int numa_node, struct rte_mempool *mp)
+{
+	int ret;
+
+	switch (ty) {
+	case QUEUE_TYPE_RX:
+		ret = rte_eth_rx_queue_setup(port_id, queue_id,
+			GATEKEEPER_NUM_RX_DESC, numa_node, NULL, mp);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT, "Failed to configure port %hhu rx_queue %hu (err=%d)!\n",
+				port_id, queue_id, ret);
+			return ret;
+		}
+		break;
+	case QUEUE_TYPE_TX:
+		ret = rte_eth_tx_queue_setup(port_id, queue_id,
+			GATEKEEPER_NUM_TX_DESC, numa_node, NULL);
+		if (ret < 0) {
+			RTE_LOG(ERR, PORT, "Failed to configure port %hhu tx_queue %hu (err=%d)!\n",
+				port_id, queue_id, ret);
+			return ret;
+		}
+		break;
+	default:
+		RTE_LOG(ERR, GATEKEEPER,
+			"Unsupported queue type (%d) passed to %s!\n",
+			ty, __func__);
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * Get a queue identifier for a given functional block instance (lcore),
+ * using a certain interface for either RX or TX.
+ */
+int
+get_queue_id(struct gatekeeper_if *iface, enum queue_type ty,
+	unsigned int lcore)
+{
+	int16_t *queues;
+	int ret;
+	uint8_t port;
+	unsigned int numa_node;
+	struct rte_mempool *mp;
+	int16_t new_queue_id;
+
+	RTE_ASSERT(lcore < RTE_MAX_LCORE);
+	RTE_ASSERT(ty < QUEUE_TYPE_MAX);
+
+	queues = ty == QUEUE_TYPE_RX ? iface->rx_queues : iface->tx_queues;
+
+	if (queues[lcore] != GATEKEEPER_QUEUE_UNALLOCATED)
+		goto queue;
+
+	/* Get next queue identifier. */
+	new_queue_id = rte_atomic16_add_return(ty == QUEUE_TYPE_RX ?
+		&iface->rx_queue_id : &iface->tx_queue_id, 1);
+	RTE_ASSERT(new_queue_id != GATEKEEPER_QUEUE_UNALLOCATED);
+	queues[lcore] = new_queue_id;
+
+	/* Configure this queue on all ports of this interface. */
+	numa_node = rte_lcore_to_socket_id(lcore);
+	mp = config.gatekeeper_pktmbuf_pool[numa_node];
+	for (port = 0; port < iface->num_ports; port++) {
```

Comment:
Let's keep the code as it is now, but adding comments to explain what you say in the previous comment would help a lot.